### PR TITLE
gathering hashes for .whl files

### PR DESCRIPTION
**Pull request by bissenbay**
thoth.python.source._download_artifacts_sha() is the place where actual .whl file download process is happening. So, we added to this function another attribute with_included_files: bool = False. 
And now in digests-fetcher inside fetch.py we can issue self.source.get_package_hashes(package_name, package_version, True) to gather hashes for internal files of .whl file. Example JSON response:
```json
[
   {
      "name": "tensorflow-1.10.0-cp27-cp27m-macosx_10_11_x86_64.whl",
      "sha256": "9eb5d117089b5cfa33c6dd32fd05574662a8167de53b686708b683e4af9bc687",
      "included_files": {
         "tensorflow-1.10.0.data/purelib/tensorflow/api_template.__init__.py": "aea7dac16d1ae0b03da0ba0efc0afda29e2841ab356a9b65df7739bb37810a90",
         "tensorflow-1.10.0.data/purelib/tensorflow/__init__.py": "85e0a47a281a3e09ea244c9632996571c46b8f15f262e89afa13d4649a9cbaef",
   }
]
```
Otherwise "included_files" attribute would have a values of false.

TO NOTE:
On line 333, I changed the code calculating hash from response.iter_content(chunk_size=1024) to hashlib.sha256(response.content) the reason is that iter_content would consume the content and I would get 'The content for this response was already consumed') RuntimeError: The content for this response was already consumed.

Examples to run:
For thoth-index: ./thoth-digests-fetcher python -p tensorflow -v 1.13 -i https://tensorflow.pypi.thoth-station.ninja/index/fedora29/1.13/jemalloc/simple
For pypi: ./thoth-digests-fetcher python -p tensorflow -v 1.10 -i https://pypi.org/simple

### CODE REVIEW

**Comment by bissenbay** [comment]
I have changes JSON output structure and that is my first suggested structure please @goern @fridex take a look and give your feedback:
```json
[{"name": "tensorflow-1.13.1-cp36-cp36m-manylinux1_x86_64.whl", "type": "directory", "sha256": "7110a2d64de1866398baab4d8ce50791071212544e2bc1562af432cd067831a3", "contents": [{"name": "tensorflow-1.13.1.dist-info", "type": "directory", "contents": [{"name": "top_level.txt", "type": "file", "sha256": "86b17d1dc4074e43210ef69c409e7421ab4eeb08a359221e3c108e500989cf35"}]}]}]
```
The other thing I want to mention is that I wanted to work only with memory data instead of saving downloaded .whl zip file to disk. However, the function at line 138 def path_to_root_folder(path) requires a directory, so I could not come up with a way where I could send a root directory inside .zip file to that function. Then I decided extract .whl file to a folder, do the hashing stuff and after deleting the folder with its contents.

**Comment by fridex** [comment]
In the provided output, please change type `directory` to `wheel` for wheel files.

> Then I decided extract .whl file to a folder, do the hashing stuff and after deleting the folder with its contents.

That sounds like a good approach :+1:

File: thoth/python/source.py [hunk 1]
```diff
@@ -125,13 +131,58 @@ def _warehouse_get_api_package_info(self, package_name: str) -> dict:
         response.raise_for_status()
         return response.json()
 
-    def _warehouse_get_package_hashes(self, package_name: str, package_version: str) -> typing.List[dict]:
+    def _construct_contents(self, path: str):
```
Comment (leonjia0112) [inline thoth/python/source.py]: It would be cool if you add `-> dict` as the other functions.

File: thoth/python/source.py [hunk 2]
```diff
@@ -323,12 +372,20 @@ def provides_package_version(self, package_name: str, package_version: str) -> b
...
+    def get_package_hashes(self, package_name: str, package_version: str, with_included_files: bool = False) -> list:
         """Get information about release hashes available in this source index."""
         if self.warehouse:
-            return self._warehouse_get_package_hashes(package_name, package_version)
+            return self._warehouse_get_package_hashes(package_name, package_version, with_included_files)
+
+        artifacts_sha = self._download_artifacts_sha(package_name, package_version, with_included_files)
+        result = []
+        for artifact_item in artifacts_sha:
+            print(len(artifact_item))
```
Comment (leonjia0112) [inline thoth/python/source.py]: Is this a extra `print`?

File: thoth/python/source.py [hunk 3]
```diff
@@ -323,12 +372,20 @@ def provides_package_version(self, package_name: str, package_version: str) -> b
...
+            doc = {}
+            doc['name'] = artifact_item[0]
+            doc['sha256'] = artifact_item[1]
+            if with_included_files:
+                doc['contents'] = artifact_item[2]
```
Comment (leonjia0112) [inline thoth/python/source.py]: I am also wondering would it be good if we could check the len before we use it?

**Comment by bissenbay** [comment]
@fridex I have made some changes according to your suggestions. Please take a look and give your feedback.
I just want to remind you that there are 2 cases. 
One when the url to download the actual .whl file contains sha256 value in which we don't have to download, we just take the value from url.
In the other case, we must download .whl file in order to calculate its hash. As you suggested I have separated _get_contents() method and now it has own logic where it takes url and does the downloading process inside the method and returns JSON it constructed. So, in this case, I cannot use _get_contents() both purposes to calculate hash for .whl and its contents because that method only cares about the contents inside. 
The only thing I could come up with was to create a separate function called _get_hash() which calculates hash for .whl file itself.
Please take a closer look at line 361 that is the place where it gets triggered. Then, right after that call there is _get_contents() which again going to download the .whl (but this time it only works with contents).
Should I try to achieve the case where there is only one download happening?

File: thoth/python/source.py (hunk 1 above)
Comment (bissenbay) [reply thoth/python/source.py]: @leonjia0112 Thanks a lot! I have added.

File: thoth/python/source.py (hunk 2 above)
Comment (bissenbay) [reply thoth/python/source.py]: Yes, those were for debugging purposes!

File: thoth/python/source.py (hunk 3 above)
Comment (bissenbay) [reply thoth/python/source.py]: @fridex what do you think? I think like if there is a flag with_included_files, then there needs to be third data which contents

**Pull request merged**